### PR TITLE
Refactor handling of GitHub token

### LINKED ISSUE #657: Where to check validity of GitHub PAT?
Look at `github_token()`, `github_user()`, `check_github_token()`. Is there a way to rationalize?

  * Why do we return `""` if no token found vs. `NULL` or erroring?
    - One answer is that this is what `gh::gh_token()` does. Which is exported only in the dev version ... hence the version here.
    - Another answer is that in some of our uses we don't absolutely need the token, i.e. we're GETting a public resource.
  * @Hadley asks "Should `github_token()` print a message (authenticating with GitHub using personal access token) and cleanly error if it can't find a token?" [here](https://github.com/r-lib/usethis/pull/656#discussion_r266192898).
    - That basically describes current behaviour of `check_github_token()`.

### PR DESCRIPTION
**Pull request by jennybc**
Closes #657 Where to check validity of GitHub PAT?

Notes to self are in the issue. Highlights:

  * Put definitions and checks closer to where they are used.
  * `""` is the PAT when we have no PAT. This is expedient, because it's what gh expects.
  *  We have two levels of PAT checking:
    - ~`validate_github_token()`~ `check_github_token()` does a rigorous check, including against the GitHub API. Use this when we know we truly need a token or git credentials.
    - ~`maybe_github_token()`~ `check_github_token(allow_empty = TRUE)` tolerates `""` but, otherwise, does full checks. Use this when we can get away with having no token or credentials, i.e. reading a public resource.

### CODE REVIEW

File: R/github.R [hunk 1]
```diff
@@ -305,17 +305,54 @@ check_no_github_repo <- function(owner, repo, host, auth_token) {
...
     ui_stop(c(
-      "No GitHub {ui_code('auth_token')}.",
-      "Provide explicitly or make available as an environment variable.",
-      "See {ui_code('browse_github_token()')} for help setting this up."
+      msg,
+      "See {ui_code('browse_github_token()')} for help storing a token as an environment variable."
     ))
   }
+
+  if (length(auth_token) != 1) {
```
Comment (hadley) [inline R/github.R]: This feels a little over-engineered to me. I think `if(!is.character(auth_token) || length(auth_token) != 1)` would be adequate.

File: R/github.R [hunk 2]
```diff
@@ -305,17 +305,54 @@ check_no_github_repo <- function(owner, repo, host, auth_token) {
...
+## tolerate "", but explicitly validate the token otherwise
+maybe_github_token <- function(auth_token = github_token()) {
+  if (!have_github_token(auth_token)) {
+    return("")
+  }
+  validate_github_token(auth_token)
+}
+
+## rigorous checks, including against the GitHub API
+validate_github_token <- function(auth_token = github_token()) {
```
Comment (hadley) [inline R/github.R]: I think this should be `check_github_token()` for consistency with the other `check_` functions (which throw an error)

File: R/github.R [hunk 3]
```diff
@@ -305,17 +305,54 @@ check_no_github_repo <- function(owner, repo, host, auth_token) {
...
+
+# github token helpers ----------------------------------------------------
+
+## minimal check: token is not ""
+have_github_token <- function(auth_token = github_token()) {
+  !isTRUE(auth_token == "")
+}
+
+## tolerate "", but explicitly validate the token otherwise
+maybe_github_token <- function(auth_token = github_token()) {
```
Comment (hadley) [inline R/github.R]: I think you should eliminate this function in favour of an additional argument, e.g. `check_github_token(allow_empty = TRUE)`

File: R/create.R [hunk 4]
```diff
@@ -240,22 +240,27 @@ check_not_nested <- function(path, name) {
   invisible()
 }
 
-rationalize_fork <- function(fork, repo_info, pat_available, user = NULL) {
-
+rationalize_fork <- function(fork, repo_info, auth_token) {
```
Comment (hadley) [inline R/create.R]: I think this function would be a little clearer if you stored `have_github_token(auth_token)` in a variable

File: R/create.R [hunk 5]
```diff
@@ -240,22 +240,27 @@ check_not_nested <- function(path, name) {
...
   owner <- repo_info$owner$login
 
   if (is.na(fork)) {
-    fork <- pat_available && !isTRUE(perms$push)
+    fork <- have_github_token(auth_token) && !isTRUE(perms$push)
   }
 
-  if (fork && !pat_available) {
-    check_github_token(auth_token = NULL)
+  if (fork && !have_github_token(auth_token)) {
```
Comment (hadley) [inline R/create.R]: I wonder if this logic would be clearer if the outer condition tested for the presence/absence of the `auth_token`

File: R/github.R [hunk 6]
```diff
@@ -305,17 +305,54 @@ check_no_github_repo <- function(owner, repo, host, auth_token) {
...
   }
+
+  if (length(auth_token) != 1) {
+    local_stop("GitHub {ui_code('auth_token')} must be a single string.")
+  }
+  if (!is_string(auth_token)) {
+    bad_class <- paste(class(auth_token), collapse = "/")
+    local_stop("GitHub {ui_code('auth_token')} must be character, not {bad_class}.")
+  }
+  if (isTRUE(auth_token == "")) {
```
Comment (hadley) [inline R/github.R]: I think `identical(auth_token, "")` expresses this intent more clearly

File: R/pr.R [hunk 7]
```diff
@@ -93,11 +93,7 @@ pr_fetch <- function(number,
   check_uses_github()
   check_uncommitted_changes()
 
-  auth_token <- github_token()
-  check_github_token(auth_token)
-
-  protocol <- github_remote_protocol()
-  credentials <- git2r_credentials(protocol, auth_token)
+  auth_token <- maybe_github_token()
```
Comment (hadley) [inline R/pr.R]: I think this should be moved to the arguments for consistency with the other functions

File: R/pr.R [hunk 8]
```diff
@@ -322,7 +320,7 @@ pr_find <- function(owner,
     owner = owner,
     repo = repo,
     .limit = Inf,
-    .token = github_token()
+    .token = maybe_github_token()
```
Comment (hadley) [inline R/pr.R]: Again, should be put in arguments, for consistency

File: R/pr.R (hunk 7 above)
Comment (jennybc) [reply R/pr.R]: Two weeks ago you said:
> I would rather not add these arguments, if the only argument is for consistency. I think we should be moving away from these arguments in general, in favour of better tooling for declaring credentials globally.

and I agree with Past Hadley 😉

https://github.com/r-lib/usethis/pull/656#discussion_r266192941

File: R/github.R (hunk 2 above)
Comment (jennybc) [reply R/github.R]: Yeah, this was a leftover from experimenting with two different levels of checks.

**Pull request merged**